### DISCUSSION TITLE
Charged and discharged Energie in HomeAssistant

### DISCUSSION THREAD

**Discussion by busaku** [reinhard-brandstaedter/solarflow-control#203] [Ideas]
Hey :)

is it possible to get this data in Homeassistant?
Or can i calculate it with the given data?

Thank you :)

**Comment by rostchri** [comment]
That is a very good point and really bad that zendure is not providing these counters out of the box. However it would be possible to calculate the charged and discharged amount of energy on your own by using influxdb/grafana/homeassistant. You need to constantly store the outputPackPower- and packInputPower-counters (see mqtt), and afterwards it is possible to use calculations explained for example in https://community.grafana.com/t/daily-power-consumption-from-watts/16196/33 to calculate appropriate kwh-counters for a specified time-range. 
But of course the solarflow-control daemon is already constantly observing mqtt-values and it would be much easier for the people (instead of calculating that on their own as explained above) if reinhard-brandstaedter would add some simple aggregation-code into the solarflow-control daemon which updates two daily kwh-counters (charged/discharged kwh per day). That would be a great and easy enhancement for solarflow-control and would compensate the zendure-lack.

**Comment by reinhard-brandstaedter** [comment]
I'm also using a timeseries database to store the current power values (pull them off mqtt via telegraf/homeassistant) and do the calculation visualization externally (either in homeassistant or via influx queries).
@rostchri is right, I could do the calculation also within solarflow-control, but since it doesn't have any persistent storage it would again need to rely on an external storage where deltas of consumption need to be stored and then again be added/calculated for a complete picture.
From start to stop of solarflow-control could of course calculate consumption and even report it periodically but as said it would always be a delta report as nothing is persisted. So I'd rather leave the long-term aggregation to the external tools.

In homassistant it's actually rather easy: use a riemann sum integral helper to get Wh from Watts: https://www.home-assistant.io/integrations/integration/

In influxdb you can also use an integral function, I'm doing it for my panels like this (which gives me the kWh produced per week):
```
from(bucket: "solar")
  |> range(start: time(v: 2022-09-01T00:00:00Z), stop: date.add(d: 1mo, to: time(v: ${__to:date:YYYY-MM}-01T00:00:00Z)))
  |> filter(fn: (r) => r["entity_id"] == "solarpanels_power")
  |> filter(fn: (r) => r["_field"] == "value")
  |> aggregateWindow(every: 1h,
          fn: (tables=<-, column) =>
            tables
              |> integral(unit: 1h)
              |> map(fn: (r) => ({ r with _value: r._value / 1000.0})))
  |> group()
  |> aggregateWindow(every: 1w, offset: -3d, fn: sum, createEmpty: false)
```

**Comment by rostchri** [comment]
Just use mqtt itself as persistent/external storage. The two counters should only represent the daily wh-counters. At midnight (or maybe at sunrise) the two counters will be set to zero, otherwise the mqtt-counters will be incremented by outputPackPower/packInputPower with a correction-factor based on your constant duty-interval. This really would help all the people who are not so experienced in using the long-term aggregation homeassistant riemann or influxdb integral functions. And these counters will be good as additional indicators how efficent the batteries are working in a typical day/night cycle in terms of how much energy goes into and how much energy comes out of the battery

**Comment by busaku** [comment]
Hmh, but (maybe iam stupid), but there is no value/entity for Riemann/Integral Helper?
Maybe i need to add 2 more helpers:

if solar > output, thant add to battery charging
and the sec. for discharging. 

and on top of these helpers i can add an integral

<img width="339" alt="image" src="https://github.com/reinhard-brandstaedter/solarflow-control/assets/25269085/2c2edf45-0aaf-4a3f-a5e5-74e0fe048e42">

or is OutputPackPower = charging and PackInputPower = discharging?

**Comment by rostchri** [comment]
outputPackPower	= battery charge [W]
packInputPower = battery discharge [W]
this mapping is correct. seems to be wrong (names are misleading and documentation in the internet is partly wrong), but i‘m using this in my calculations for months and this is working reasonable

**Comment by busaku** [comment]
Thanks, that was easy :)
